src/charm.py

```diff
@@ -24,36 +29,81 @@ def __init__(self, *args: typing.Any):
-        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
 
     def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
-        """Handle changed configuration."""
-        self.unit.status = ops.ActiveStatus()
+        """Handle config changed."""
+        self.reconcile()
 
     def _on_install(self, _: ops.InstallEvent) -> None:
         """Handle install."""
-        self.unit.status = ops.MaintenanceStatus("Preparing irc bridge")
-        self.irc.prepare()
-        self.unit.status = ops.ActiveStatus()
+        self.reconcile()
+
+    def _on_upgrade_charm(self, _: ops.UpgradeCharmEvent) -> None:
+        """Handle upgrade charm."""
+        self.reconcile()
 
     def _on_start(self, _: ops.StartEvent) -> None:
         """Handle start."""
-        self.irc.start()
-        self.unit.status = ops.ActiveStatus()
+        self.reconcile()
 
     def _on_stop(self, _: ops.StopEvent) -> None:
         """Handle stop."""
-        self.irc.stop()
+        self.unit.status = ops.MaintenanceStatus("Stopping charm")
+        self._irc.stop()
 
-    def _on_upgrade_charm(self, _: ops.UpgradeCharmEvent) -> None:
-        """Handle upgrade-charm."""
-        self.unit.status = ops.MaintenanceStatus("Upgrading dependencies")
-        self.irc.prepare()
+    def _charm_config(self) -> CharmConfig:
+        """Reconcile the charm.
+
+        Returns:
+            CharmConfig: The reconciled charm configuration.
+        """
+        return CharmConfig(
+            ident_enabled=self.model.config.get("ident_enabled", None),
+            bot_nickname=self.model.config.get("bot_nickname", None),
+            bridge_admins=self.model.config.get("bridge_admins", None),
+        )
+
+    def reconcile(self) -> None:
+        """Reconcile the charm.
+
+        This is a more simple approach to reconciliation,
+        adapted from Charming Complexity sans state and observers.
+
+        Being a simple charm, we don't need to do much here.
+
+        Ensure we have a database relation,
+        ensure we have a relation to matrix,
+        populate database connection string and matrix homeserver URL
+        in the config template and (re)start the service.
+        """
+        ops.MaintenanceStatus("Reconciling charm")
+        try:
+            logger.info("DB Reconciling charm")
+            db = self._database.get_db()
+            if db is None:
+                self.unit.status = ops.BlockedStatus("Database relation not found")
+                return
+        except ValidationError:
+            self.unit.status = ops.MaintenanceStatus("Database configuration not correct")
+            return
+        logger.info("Matrix Reconciling charm")
+        matrix = self._matrix.reconcile()
+        try:
+            logger.info("Config Reconciling charm")
+            config = self._charm_config
+        except ValidationError as e:
+            self.unit.status = ops.BlockedStatus(f"Invalid configuration: {e}")
```

Comment:
Shouldn't we log the exception because it might not fit in the status message.

Reply:
Good point, I'll add in a logger.exception.